### PR TITLE
Add simple end-to-end example of DP training of a transformer

### PR DESCRIPTION
**Pull request by moksha-hub**
This addresses issue #86 by adding a single-file, forkable example that uses JAX Privacy core library (not Keras API), implements a minimal transformer for next-token prediction, demonstrates per-example gradient clipping, noise calibration from privacy budget, and gaussian_privatizer for noise addition.

### CODE REVIEW

**Comment by moksha-hub** [comment]
@googlebot I signed it!

File: examples/dp_sgd_transformer_example.py [hunk 1]
```diff
@@ -0,0 +1,399 @@
...
+# Transformer hyperparameters
+VOCAB_SIZE = 100
+SEQ_LEN = 16
+D_MODEL = 32
+NUM_HEADS = 2
+D_FF = 64
+NUM_LAYERS = 2
+
+
+def init_transformer_params(key):
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_example.py:66]: You might want to add a @jax.jit annotation here

File: examples/dp_sgd_transformer_example.py [hunk 2]
```diff
@@ -0,0 +1,399 @@
...
+
+
+def layer_norm(x, scale, bias, eps=1e-5):
+  """Apply layer normalization."""
+  mean = jnp.mean(x, axis=-1, keepdims=True)
+  var = jnp.var(x, axis=-1, keepdims=True)
+  return scale * (x - mean) / jnp.sqrt(var + eps) + bias
+
+
+def attention(x, wq, wk, wv, wo):
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_example.py:135]: Maybe consider implementing this in terms of Jax.nn.dot_product_attention

File: examples/dp_sgd_transformer_example.py [hunk 3]
```diff
@@ -0,0 +1,399 @@
...
+    # [START DP SETUP]
+    print("\nSetting up differential privacy...")
+    
+    # Calculate noise multiplier from privacy budget
+    accountant = analysis.DpsgdTrainingAccountant(
+        dp_accountant_config=accountants.PldAccountantConfig()
+    )
+    num_updates = num_epochs * (train_size // batch_size)
+    
+    noise_multiplier = calibrate.calibrate_noise_multiplier(
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_example.py:301]: Add a comment that this API calibrates noise to the mean gradient rather than the sum of gradients.

File: examples/dp_sgd_transformer_example.py [hunk 4]
```diff
@@ -0,0 +1,399 @@
...
+      loss = aux_outputs.values.mean()
+      
+      # Average gradients over batch
+      mean_grads = jax.tree.map(lambda x: x / batch_size, grads)
+      
+      # Add calibrated noise
+      noisy_grads, noise_state = privatizer.update(mean_grads, noise_state)
+      
+      # SGD update
+      params = jax.tree.map(
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_example.py]: Maybe write this in terms of an optax.GradientTransformation (default to adamw), then use optax.apply_upates here

File: examples/dp_sgd_transformer_example.py [hunk 5]
```diff
@@ -0,0 +1,399 @@
...
+          lambda p, g: p - learning_rate * g, params, grads
+      )
+      return params, loss
+  
+  # Training loop
+  print(f"\n{'=' * 60}")
+  print(f"Training {'with DP' if use_dp else 'without DP'}...")
+  print(f"{'=' * 60}")
+  
+  for epoch in range(num_epochs):
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_example.py:372]: Can you use a jax_privacy.batch_selection here?  The accounting assumes Poisson sampling, not shuffling

**Comment by moksha-hub** [comment]
@ryan112358 now can you verify again Thanks for previous reply.

File: examples/dp_sgd_transformer_example.py [hunk 6]
```diff
@@ -357,7 +367,11 @@ def train_step(params, batch_input, batch_target):
   print(f"{'=' * 60}")
   
   for epoch in range(num_epochs):
-    # Shuffle data
+    # Note: For strict privacy accounting, you should use Poisson sampling
+    # (each example included independently with probability batch_size/train_size)
+    # rather than shuffling. For simplicity, we use shuffling here, but in
+    # production you should use jax_privacy.batch_selection for proper sampling.
+    # See: https://github.com/google-deepmind/jax_privacy for batch_selection API.
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_example.py]: In this case since the inputs are just arrays, you should be able to index into them with Poisson sampled indices without issue, so might as well do that here

File: examples/dp_sgd_transformer_example.py [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+  # Gather log probs for target tokens
+  batch_size, seq_len = target_ids.shape
+  target_log_probs = jnp.take_along_axis(
+      log_probs, target_ids[:, :, None], axis=-1
+  ).squeeze(-1)
+  
+  return -jnp.mean(target_log_probs)
+
+
+def generate_synthetic_data(key, num_samples):
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_example.py]: Is there any performance difference between using numpy vs. Jax for this function?  Numpy might be preferable here as a way to ensure that the full dataset lives on CPU, and only batches get transferred to GPU via Jax.  Your train_step should handle this automatically

File: examples/dp_sgd_transformer_example.py [hunk 8]
```diff
@@ -0,0 +1,413 @@
...
+      updates, opt_state = optimizer.update(noisy_grads, opt_state, params)
+      params = optax.apply_updates(params, updates)
+      return params, opt_state, loss, noise_state
+    
+  else:
+    @jax.jit
+    def train_step(params, batch_input, batch_target):
+      """Single non-private training step."""
+      loss, grads = jax.value_and_grad(loss_fn)(params, batch_input, batch_target)
+      params = jax.tree.map(
```
Comment (ryan112358) [inline examples/dp_sgd_transformer_example.py:362]: This should also use optax

File: examples/dp_sgd_transformer_example.py (hunk 6 above)
Comment (ryan112358) [reply examples/dp_sgd_transformer_example.py]: Please resolve.  You can use jax_privacy.batch_selection.CyclicPoissonSampling to generate batch_idx in a DP-compliant manner.
